src/System.IO.FileSystem/src/System/IO/Win32FileStream.cs

```diff
@@ -468,10 +469,13 @@ protected override void Dispose(bool disposing)
                 _canRead = false;
                 _canWrite = false;
                 _canSeek = false;
-                // Don't set the buffer to null, to avoid a NullReferenceException
-                // when users have a race condition in their code (ie, they call
-                // Close when calling another method on Stream like Read).
-                //_buffer = null;
+
+                if (_buffer != null)
+                {
+                    ArrayPool<byte>.Shared.Return(_buffer, clearArray: true);
```

Comment:
re: (4) if there was another Api where the user provided the buffer; and thus the onus was on them to also free it? Then programming error would be more valid.

Also would allow alternative pools